File: alibaba-rsocket/src/main/kotlin/org/microjservice/rsocket/alibaba/CloudEventToListenerConsumer.java
package org.microjservice.rsocket.alibaba;

import com.alibaba.rsocket.cloudevents.CloudEventImpl;
import com.alibaba.rsocket.events.CloudEventsConsumer;
import io.micronaut.context.event.ApplicationEventPublisher;

import reactor.core.publisher.Mono;

/**
 * CloudEvent to @EventListener handler
 *
 * @author leijuan
 */
public class CloudEventToListenerConsumer implements CloudEventsConsumer {
    private ApplicationEventPublisher eventPublisher;

    @Override
    public boolean shouldAccept(CloudEventImpl<?> cloudEvent) {
        return true;
    }

    @Override
    public Mono<Void> accept(CloudEventImpl<?> cloudEvent) {
        return Mono.fromRunnable(() -> {
            eventPublisher.publishEvent(cloudEvent);
        });
    }
}

File: alibaba-rsocket/src/main/kotlin/org/microjservice/rsocket/alibaba/CloudEventToListenerConsumer.kt
package org.microjservice.rsocket.alibaba

import com.alibaba.rsocket.events.CloudEventsConsumer
import io.micronaut.context.event.ApplicationEventPublisher
import com.alibaba.rsocket.cloudevents.CloudEventImpl
import reactor.core.publisher.Mono
import java.lang.Void

/**
 * CloudEvent to @EventListener handler
 *
 * @author leijuan
 */
class CloudEventToListenerConsumer(private val eventPublisher: ApplicationEventPublisher<CloudEventImpl<*>>) :
    CloudEventsConsumer {

    override fun shouldAccept(cloudEvent: CloudEventImpl<*>) = true

    override fun accept(cloudEvent: CloudEventImpl<*>): Mono<Void> {
        return Mono.fromRunnable { eventPublisher.publishEvent(cloudEvent) }
    }

}